Reworks the migration sections to better showcase Hypershift (#2230)
timescale/docs: use-timescale/migration/hypershift-config.md

excerpt: Migrate an existing PostgreSQL database to Timescale Cloud in a single step
products: [cloud]
keywords: [data migration, Hypershift]
tags: [ingest, Hypershift, postgresql]
---

# Hypershift configuration file

Hypershift uses a YAML configuration file to determine how to set up your new
Timescale database. You can create your own file, or use the example file as a
starting point. To complete your file, you need these details for the tables
that you want to convert to hypertables:

*   The schema that contains the table
*   The name of the table
*   The name of the `time` column of that table
*   The chunk time interval to use
*   The compression policy you want to use

If you are not sure what chunk time interval to use, see the
[time partitioning section][chunk-time].

Use this format:

```yml
source: '<SOURCE_DB_URI>'
target: '<TARGET_DU_URI>'
actions: ['clone', 'verify']
include_tables: []
exclude_tables: []
include_schemas: []
exclude_schemas: []

hypertable_configs:
  name: <table_name>
  schema: <schema_name>
  time_column_name: <time_column_name>
  chunk_time_interval: 7d
  compress:
    after: <in units of time>
    segmentby:
<column list>
    orderby:

> <column list>

Reviewer: We should also mention the direction field, as mentioned [here](https://github.com/timescale/hypershift/blob/1c11cf25001d69606bb79ea860b648c56f5cf8ae/src/config/mod.rs#L153)

```suggestion
- <column> <direction>
```

---
title: Configuring Hypershift
excerpt: Migrate an existing PostgreSQL database to Timescale Cloud in a single step
products: [cloud]
keywords: [data migration, Hypershift]
tags: [ingest, Hypershift, postgresql]
---

# Hypershift configuration file

Hypershift uses a YAML configuration file to determine how to set up your new
Timescale database. You can create your own file, or use the example file as a
starting point. To complete your file, you need these details for the tables
that you want to convert to hypertables:

*   The schema that contains the table
*   The name of the table
*   The name of the `time` column of that table
*   The chunk time interval to use
*   The compression policy you want to use

If you are not sure what chunk time interval to use, see the
[time partitioning section][chunk-time].

Use this format:

```yml
source: '<SOURCE_DB_URI>'
target: '<TARGET_DU_URI>'
actions: ['clone', 'verify']
include_tables: []
exclude_tables: []
include_schemas: []
exclude_schemas: []

hypertable_configs:
  name: <table_name>
  schema: <schema_name>
  time_column_name: <time_column_name>
  chunk_time_interval: 7d
  compress:
    after: <in units of time>

>     segmentby:
> <column list>
>     orderby:
> - <column> <direction>

Reviewer 2: Indentation is incorrect:
```suggestion
    segmentby:
      - <column>
    orderby:
      - <column> <direction>
```

```

## Source and target database details

This is the connection string to use for the databases. The source is the
database you are migrating from, and the target is the new database that you are
migrating to. Connection strings usually contain the hostname to access the
database, the name of the database, the port the database can be accessed on,
and username and password details to log in to the database. If you do not want
to provide this information in the configuration file, you can provide these
details on the command line instead. For more information, see the

> [hypershift command line options][hypershift-cli].

Reviewer 2: This link is broken

Connection strings are in the format:

```bash

> <USERNAME>:<DB_NAME>@<HOSTNAME>:<PORT>

Reviewer 2: ```suggestion
postgres://<USERNAME>:<PASSWORD>@<HOSTNAME>:<PORT>/<DB_NAME>
```

```

For example:

```bash

> tsdbadmin:my_database@timescale-cloud.com:30001

Reviewer 2: ```suggestion
postgres://tsdbadmin:my_password@timescale-cloud.com:30001/my_database
```

```

## Actions

The available options for hypershift are `clone` and `verify`.

### Clone

Use the `clone` action to perform a hypershift migration.

### Verify

> Use the `verify` action to check if a migration is going to work successfully,
> without performing the migration itself.

Reviewer 2: ```suggestion
Use the `verify` action to check if the migration was successful. This does a rough
check over the data in the tables on the source and target database to ensure that
they contain the same values.
```

## Include and exclude tables

If you want to include or exclude tables from the migration, you can use the
`include_tables` and `exclude_tables` parameters.

Use `include_tables` to declare the tables that should be migrated. For example:

```yaml
include_tables:
- '"MixedCaseSchema"."MixedCaseTable"'
- public.my_table
- public.a_small_table
- other_schema.table
```

Use `exclude_tables` to declare tables that should not be migrated. For example:

```yaml
exclude_tables:
- '"MixedCaseSchema"."MixedCaseTable"'
- public.my_table
- public.a_small_table
- other_schema.table
```

If you declare the `include_tables` parameter, you cannot also declare the
`exclude_tables` parameter. Similarly, if you declare the `exclude_tables`
parameter, you cannot also declare the `include_tables` parameter.

### Dependent objects for tables

> It is possible for a table to depend on other object, for example the schema

Reviewer 2: ```suggestion
It is possible for a table to depend on other objects, for example the schema
```

*   `deferred`: Check all foreign keys in a batch after the migration is finished.
*   `unchecked`: Leave all foreign keys unchecked.
*   `disabled`: Disable foreign key checking entirely.

This parameter defaults to `unchecked`.

### Logging format

Use the `log_format` parameter to control which file format to save log files
in. The options for this parameter are:

*   `json`: Save logs in `json` format.
*   `text`: Save logs as plain text.

This parameter defaults to `text`.

### Create the target database

Use the `create_db` parameter to control whether or not hypershift creates a new
target database during the migration. The options for this parameter are:

*   `true`: Create a new target database during migration.
*   `false`: Do not create a new target database during migration.

This parameter defaults to `false`.

### Drop incompatible hypertable constraints

In some cases, tables can have constraints on them that can't be retained when
they are converted to a hypertable. For example, if a table has a foreign key
that points to a hypertable, the foreign key can't be retained, and an error is
shown. When this occurs, the migrations fails. You can use the

> `drop_incompatible_hypertable_contraints` parameter to silently drop any

Reviewer 2: ```suggestion
`drop_incompatible_hypertable_constraints` parameter to silently drop any
```

hypertable_configs: [<hypertable_configuration>] # Default = []
# Minimum size of the table (in MiB) for which parallel
# copying of data should be done.
parallel_min_table_size: <u16> # Default = 128
# Specify when should foreign keys be checked.
foreign_keys_priority: <VALID: immediate, deferred, unchecked, disabled> # Default = unchecked
# Specify the logging format to be used.
log_format: <VALID: json, text> # Default = text
# Create the target database.
create_db: <boolean> # Default = false
# Drop constraints that prevent hypertables from being created.
#
# Some constraints like foreign keys pointing to a Hypertable cannot be
# retained when migrating and will throw an error.
#
# Use this option to not create those constraints instead.
drop_incompatible_hypertable_constraints: <boolean> # Default = false
# Hide migration progress bar.
hide_progress: <boolean> # Default = false
# Display verbose logs with progress bar.
verbose: <boolean> # Default = false
# Do not collect anonymous telemetry during migration. By default,
# Hypershift collects anonymous statistics that are important for
# its improvement.
no_telemetry: <boolean> # Default = false
```

</Collapsible>

> [chunk-time]: /use-timescale/:currentVersion:/hypertables/about-hypertables#best-practices-for-time-partitioning

Author: ```suggestion
[chunk-time]: /use-timescale/:currentVersion:/hypertables/about-hypertables#best-practices-for-time-partitioning
[hypershift-cli]:  /use-timescale/:currentVersion:/migration/about-hypershift/
```
